### PR TITLE
Add more intelligent caching logic to JPX movies for issue #34

### PR DESCRIPTION
**Pull request by dgarciabriseno**

### CODE REVIEW

**Comment by dgarciabriseno** [comment]
New logic is implemented so when a JPX image is requested, the summary now has a new field "cache"
When cache = True, the existing JPX is used.
The cache value is determined by the requested end time. If end time is in the future, then we don't cache the jpx image so that when you request it again, it will generate a new jpx with newer frames.

**Comment by bogdanni** [comment]
It should not be about the future, it should be:

1. Read if possible the existing JSON file and fetch the timestamps therein.
2. If 1. succeeds and if current timestamps are the same with the JSON timestamps, use cached JPX & JSON file.

**Comment by dgarciabriseno** [comment]
To rephrase, do you mean something like this?
1. Read (if possible) the existing JSON file and store timestamps from the json file.
2. Check database for images within the time range and compare database timestamps to the json timestamps.

**Comment by bogdanni** [comment]
Yes.
There can be new files in the database, for example due to a synchronization after a data provider outage.

**Comment by dgarciabriseno** [comment]
@bogdanni Okay, I updated the logic so each request will query the database for new frames and compare these timestamps against the summary file.
If they match, then it will use the movie on-disk. If not, then it will continue to run kdu_merge to generate the new movie.

File: src/Image/JPEG2000/HelioviewerJPXImage.php [hunk 1]
```diff
@@ -36,6 +36,9 @@ class Image_JPEG2000_HelioviewerJPXImage extends Image_JPEG2000_JPXImage {
      * @param bool   $linked     Whether or not requested JPX image should be
      *                           a linked JPX
      * @param string $outputFile Filename to use for generated JPX file
+     * 
+     * @param int    $currentTime An epoch timestamp to override current system
```
Comment (dgarciabriseno) [inline src/Image/JPEG2000/HelioviewerJPXImage.php:40]: Remove this comment.

File: src/Image/JPEG2000/HelioviewerJPXImage.php [hunk 2]
```diff
@@ -110,7 +110,64 @@ public function __construct($sourceId, $startTime, $endTime, $cadence, $linked,
...
+    /**
+     * Loads and returns the summary contents as a json object.
+     */
+    private function _loadSummary() {
+        $summary_fp = fopen($this->_summaryFile, 'r');
+        $summary_contents = fread($summary_fp, filesize($this->_summaryFile));
+        fclose($summary_fp);
+
+        return json_decode($summary_contents);
+    }
```
Comment (dgarciabriseno) [inline src/Image/JPEG2000/HelioviewerJPXImage.php:170]: This function is no longer necessary

File: src/Image/JPEG2000/HelioviewerJPXImage.php [hunk 3]
```diff
@@ -483,5 +540,19 @@ public function printJSON($jpip, $verbose) {
...
+    public function getSummaryFile() {
+        return $this->_summaryFile;
+    }
+
+    /**
+     * Returns the path to this jpx's binary file.
+     */
+    public function getJpxFile() {
+        return $this->_jpxFile;
+    }
```
Comment (dgarciabriseno) [inline src/Image/JPEG2000/HelioviewerJPXImage.php:556]: These methods are no longer necessary

File: tests/unit_tests/HelioviewerJPXImageTest.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    private function _setupTestFiles() {
+        copy(self::TEST_JPX_FILE, self::MOVIE_DIR . self::TEST_JPX_NAME);
+        copy(self::TEST_JSON_FILE, self::MOVIE_DIR . self::TEST_JSON_NAME);
+    }
+
+    private function _cleanupTestFiles() {
+        unlink(self::MOVIE_DIR . self::TEST_JPX_NAME);
+        unlink(self::MOVIE_DIR . self::TEST_JSON_NAME);
+    }
+}
```
Comment (dgarciabriseno) [inline tests/unit_tests/HelioviewerJPXImageTest.php:91]: add newline

File: tests/unit_tests/config.php [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+define("HV_DB_PASS", "helioviewer");
+
+define("HV_ROOT_DIR", "../../docroot");
+define("HV_JP2_DIR", "/var/www/jp2");
+define("HV_JP2_ROOT_URL", "127.0.0.1"); 
+
+define("HV_MAX_JPX_FRAMES", 1000);
+define("HV_KDU_MERGE_BIN", "/usr/local/bin/kdu_merge");
+
+define("HV_TEST_ASSETS_DIR", "test_assets");
```
Comment (dgarciabriseno) [inline tests/unit_tests/config.php:15]: add newline

**Comment by bogdanni** [comment]
Is it absolutely certain that if JSON exists and JPX does not (kduMerge still busy), that the JSON is not yet sent?

**Comment by dgarciabriseno** [comment]
This condition where JPX exists and JSON does not will only happen if the server gets two simultaneous requests.

When getJPX is called, the server will first execute KDU merge, then upon completion it will create the JSON file and send the result back to the caller.

While this is happening, another request may come in and see the JPX already exists, but there is no JSON because the previous request is still being processed. This request will not generate a new JPX, instead it will wait for kdu_merge to complete, the first request will create the JSON file, and then the second request will return the same result.

**Comment by bogdanni** [comment]
OK, I'm too tired to think, I trust you :)

It's only that I've seen a couple of instances where there was a JSON, it was sent to JHV and when JHV requested the JPX, the JPIP server responded something to the effect of "file not found" (this has to be seen in the server logs). Maybe your changes prevent this.

**Comment by dgarciabriseno** [comment]
That is strange, next time this happens save the file name and I can look into it.
The JPX is created first, so I'm not sure how the JSON can be returned without it. Unless kdu_merge somehow fails silently.

We're working on bringing up a beta server to stage changes like this before pushing them to helioviewer.org, details will be announced on the blog once it is ready.

**Pull request merged**